hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/kafka/SqlAvroTest.java

```diff
@@ -33,131 +44,422 @@
+
+    private SqlMapping mapping;
+    private Schema keySchema;
+    private Schema valueSchema;
+    private Map<String, String> consumerProperties;
+
     @BeforeClass
     public static void initialize() throws Exception {
         createSchemaRegistry();
     }
 
-    private static SqlMapping kafkaMapping(String name) {
-        return new SqlMapping(name, KafkaSqlConnector.TYPE_NAME).options(
-                OPTION_KEY_FORMAT, AVRO_FORMAT,
-                OPTION_VALUE_FORMAT, AVRO_FORMAT,
-                "bootstrap.servers", kafkaTestSupport.getBrokerConnectionString(),
-                "schema.registry.url", kafkaTestSupport.getSchemaRegistryURI(),
-                "auto.offset.reset", "earliest"
-        );
+    private SqlMapping kafkaMapping(String name) {
+        return kafkaMapping(name, null, null);
+    }
+
+    private SqlMapping kafkaMapping(String name, Schema keySchema, Schema valueSchema) {
+        this.keySchema = keySchema;
+        this.valueSchema = valueSchema;
+        consumerProperties = useSchemaRegistry
+                ? ImmutableMap.of("schema.registry.url", kafkaTestSupport.getSchemaRegistryURI().toString())
+                : ImmutableMap.of(OPTION_KEY_AVRO_SCHEMA, keySchema.toString(),
+                                  OPTION_VALUE_AVRO_SCHEMA, valueSchema.toString());
+
+        return mapping = new SqlMapping(name, KafkaSqlConnector.TYPE_NAME)
+                .options(OPTION_KEY_FORMAT, AVRO_FORMAT,
+                         OPTION_VALUE_FORMAT, AVRO_FORMAT,
+                         "bootstrap.servers", kafkaTestSupport.getBrokerConnectionString(),
+                         "auto.offset.reset", "earliest")
+                .optionsIf(useSchemaRegistry,
+                           "schema.registry.url", kafkaTestSupport.getSchemaRegistryURI())
+                .optionsIf(!useSchemaRegistry,
+                           OPTION_KEY_AVRO_SCHEMA, keySchema,
+                           OPTION_VALUE_AVRO_SCHEMA, valueSchema,
+                           "key.serializer", HazelcastAvroSerializer.class.getCanonicalName(),
+                           "key.deserializer", HazelcastAvroDeserializer.class.getCanonicalName(),
+                           "value.serializer", HazelcastAvroSerializer.class.getCanonicalName(),
+                           "value.deserializer", HazelcastAvroDeserializer.class.getCanonicalName());
+    }
+
+    @Test
+    public void when_inlineSchemaUsedWithSchemaRegistry_then_fail() {
+        assumeTrue(useSchemaRegistry);
+        assertThatThrownBy(() ->
+                    kafkaMapping("kafka")
+                            .fields("id INT EXTERNAL NAME \"__key.id\"",
+                                    "name VARCHAR")
+                            .options(OPTION_VALUE_AVRO_SCHEMA, NAME_SCHEMA)
+                            .create())
+                .hasMessage("Inline schema cannot be used with schema registry");
+    }
+
+    @Test
+    public void test_schemaValidation() {
```

Comment:
IMO it would be better to split this test to separate methods for both features that are tested. For "Test non-record schema" we should check all `Schema.Type` values except RECORD.
And even better, we should have such unit tests also in `KvMetadataAvroResolverTest`

Reply:
Because of my tendency to prototyping (instead of analyzing), I usually end up writing end-to-end tests. I realized the existence of `KvMetadataAvroResolverTest` after you commented and couldn't figure out how adding unit tests will contribute to test coverage in this case.

Reply:
prototyping is fine, but after after we learn the problem structure etc. refactoring is necessary. Without refactoring we would end up with big ball of prototype code wired in strange ways (general remark, does not apply specifically do this PR which is fine with regard to design).

In unit test you might test more cases and in more direct way. If entire test setup (registry, mapping etc) is just to invoke single method (eg. schema validation) and assert that exception is thrown, it would be much more straightforward to do that in unit test instead.